lazy_load_symbols/load_sym_lazily.py

```diff
@@ -0,0 +1,171 @@
+from __future__ import absolute_import
+import re
+from subprocess import Popen, PIPE
+import gdb
+
+
+def parse_raw_symlist(syml):
+    symre = r' ([^ ]*)$'
+    symc = re.compile(symre)
+    pl = []
+    for line in syml.split('\n'):
+        m = symc.search(line)
+        if m:
+            pl.append(m.group(1))
+    return pl
+
+
+def extract_symbols_from_raw_list(rs):
+    '''
+    Saves the entire symbol names as a single long string
+    so we can say something like:
+        if symbol in symbol_list
+    '''
+    pl = ''
+    for sym in rs:
+        filth = Popen(['c++filt', sym], stdout=PIPE)
+        (sym_name, err) = filth.communicate()
+        assert not err, 'c++filt failed {}'.format(err)
+        assert sym_name, 'c++filt didn\'t demangle {}'.format(sym)
+        ec = filth.wait()
+        assert not ec, 'c++filt exited with {}'.format(ec)
+        pl += sym_name
+    return pl
+
+
+class LoadLazySymbols(gdb.Command):
+    '''
+    A command to load a recording without loading
+    the symbols and then loading only the required symbols.
+    '''
+
+    def __init__(self):
+        super(LoadLazySymbols, self).__init__('loadlazy', gdb.COMPLETE_EXPRESSION)
+        self.solist = []
+        self.sodict = {}
+
+    def load_so_list(self):
+        self.solist = gdb.execute('info sharedlibrary', to_string=True)
+        solibre = r'0x([0-9a-f]+)\s+0x([0-9a-f]+).+ ([^\0]+)\/(.+)'
+        solibc = re.compile(solibre)
+        for so in self.solist.split('\n'):
+            m = solibc.match(so)
+            if m:
+                name = m.group(4)
+                so_path = m.group(3)
+                sa = m.group(1)
+                ea = m.group(2)
+                if name not in self.sodict:
+                    self.sodict[name] = {}
+                    self.sodict[name]['symbols'] = {}
+                    self.sodict[name]['fullname'] = so_path + '/' + name
+                    self.sodict[name]['start'] = sa
+                    self.sodict[name]['end'] = ea
+                    self.sodict[name]['loaded'] = False
+                else:
+                    # this shouldn't happen but we can just verify it is indeed the same library
+                    assert self.sodict[name]['start'] == sa, 'found same' \
+                        ' library ({}) at 2 different locations: {} != {}' \
+                        .format(name, sa, self.sodict[name]['start'])
+                    assert self.sodict[name]['end'] == ea, 'found same library' \
+                        ' ({}) ending at 2 different locations: {} != {}' \
+                        .format(name, ea, self.sodict[name]['end'])
+        return self.sodict
+
+    def load_all_symbols(self):
+        for so in self.sodict:
```

Comment:
I think it's nicer to have `for so, info in self.sodict.items(): ...` rather then just get the key and use it to access values.
Or, as you don't use `so`, just do `for info in self.sodict.values(): ...`.

Reply:
done